source/cgal/intersections.cc

```diff
@@ -0,0 +1,737 @@
+#  else
+
+      Assert(
+        false,
+        ExcMessage(
+          "This function requires a version of CGAL greater or equal than 5.1.5."));
+      (void)first_simplex;
+      (void)second_simplex;
+      return {};
+#  endif
+    }
+  } // namespace internal
+
+
+
+  // Specialization for quads
+  template <>
+  std::vector<std::array<Point<2>, 3>>
+  compute_intersection_of_cells<2, 2, 2>(
+    const typename Triangulation<2, 2>::cell_iterator &cell0,
+    const typename Triangulation<2, 2>::cell_iterator &cell1,
+    const Mapping<2, 2> &                              mapping0,
+    const Mapping<2, 2> &                              mapping1,
+    const double                                       tol)
+  {
+    std::array<Point<2>, 4> vertices0, vertices1;
+    std::copy_n(mapping0.get_vertices(cell0).begin(), 4, vertices0.begin());
+    std::copy_n(mapping1.get_vertices(cell1).begin(), 4, vertices1.begin());
+
+    std::swap(vertices0[2], vertices0[3]);
+    std::swap(vertices1[2], vertices1[3]);
+    const auto intersection_test =
+      internal::compute_intersection(vertices0, vertices1);
+
+    if (!intersection_test.empty())
+      {
+        const auto &       poly      = intersection_test[0].outer_boundary();
+        const unsigned int size_poly = poly.size();
+        if (size_poly == 3)
+          {
+            // intersection is a triangle itself, so directly return its
+            // vertices.
+            return {
+              {{CGALWrappers::cgal_point_to_dealii_point<2>(poly.vertex(0)),
+                CGALWrappers::cgal_point_to_dealii_point<2>(poly.vertex(1)),
+                CGALWrappers::cgal_point_to_dealii_point<2>(poly.vertex(2))}}};
+          }
+        else if (size_poly >= 4)
+          {
+            // intersection is a polygon, need to triangulate it.
+            std::vector<std::array<Point<2>, 3>> collection;
+
+            CDT cdt;
+            cdt.insert_constraint(poly.vertices_begin(),
+                                  poly.vertices_end(),
+                                  true);
+
+            internal::mark_domains(cdt);
+            std::array<Point<2>, 3> vertices;
+
+            for (Face_handle f : cdt.finite_face_handles())
+              {
+                if (f->info().in_domain() &&
+                    CGAL::to_double(cdt.triangle(f).area()) > tol)
+                  {
+                    collection.push_back(
+                      {{CGALWrappers::cgal_point_to_dealii_point<2>(
+                          cdt.triangle(f).vertex(0)),
+                        CGALWrappers::cgal_point_to_dealii_point<2>(
+                          cdt.triangle(f).vertex(1)),
+                        CGALWrappers::cgal_point_to_dealii_point<2>(
+                          cdt.triangle(f).vertex(2))}});
+                  }
+              }
+            return collection;
+          }
+        else
+          {
+            Assert(false, ExcMessage("The polygon is degenerate."));
```

Comment:
I am not sure. If we really need an assert here. This implies that the function can only be used if there is an intersection. But how does one know that there is an intersection?

Reply:
You're right. The use case I had in mind is the one where the user knows there's an intersection (thanks to rtrees of bboxes, for instance). However, I think it makes more sense (also mathematically) to return an empty vector if two cells are not intersecting, so I removed the asserts from the code. I added this case to the tests.